### PR TITLE
feat: Inkling-Small 276B-A12B support (4-bit MLX, SSD-streamed experts)

### PR DESCRIPTION
**Pull request by NeelM0906**
Adds the fourth model family: **Inkling-Small 276B-A12B** (Thinking Machines, Apache 2.0), running from [pipenetwork/Inkling-Small-MLX-4bit](https://huggingface.co/pipenetwork/Inkling-Small-MLX-4bit) with SSD-streamed experts on a 24 GB machine.

```
$ MferenceCLI --model inklingsmall.gturbo --prompt "The capital of France is" --max-new 16 --temperature 0
 Paris. The capital of Germany is Berlin. The capital of Italy is Rome.

chat: "What is the capital of France? Answer in one word." → Paris   [stop=endOfTurn]
```

## Quality vs upstream reference

Teacher-forced against the checkpoint's own `inkling_mlx` implementation (77-token text, all positions): **mean KLD 0.0089** (median 0.0015, p95 0.039), **top-1 agreement 97.4%**, teacher-forced **PPL 9.50 vs 9.33** reference (+1.8%, FP16 intermediate noise). Per-layer hidden states match the reference at cos ≥ 0.999.

## Checkpoint selection

Of ~a dozen public MLX conversions, this is the only one whose quantization matches the engine's contract (affine 4-bit g64 with BF16 scales for embedding/attention/experts; BF16 router like DSV4). The 3-bit/2-bit builds leave attention+embeddings in BF16, which the engine has no resident path for — and the 4-bit's ~3.4 GB resident set is what fits a 24 GB box. Source pinned by revision `9d6e4720` + index SHA-256. Full analysis in `docs/INKLING_SMALL.md`.

## Architecture (all new to the engine)

- **No RoPE** — a learned relative-position bias (`wr_du` → per-head 16-dim states mixing a bias-vs-distance bank), extent per layer kind (512 local / 1024 global), log-`tau` scaling on global layers past 128k
- **Depthwise short convolutions** (K=4) at four sites per layer with streaming fp32 state
- **Sigmoid router with shared-expert sinks**: 258 logits, correction bias for selection only, softmax over logsigmoid of the 6 selected + 2 shared logits — all fp32
- **2 leading dense-FFN layers** (16 384 wide); 40 MoE layers of 256 experts top-6 riding the existing INT4 MoE phase kernels (padded to the 8-slot contract with weight-0 duplicates — cache hits, zero extra I/O)
- **BF16-ranged numerics**: FP32 residual stream (hidden legitimately reaches ~5e5; FP16 overflowed at L23), FP32 sublayer deltas, and an exact ÷32/×32 prescale through the FFN output path

## Prefill: layer-major with expert-major streaming

Prefill cost is dominated by cold expert I/O, so chunks advance one layer at a time and each unique routed expert in the chunk is **read once and applied to every token routed to it** (single-expert GLU + weighted FP32 accumulation; the next expert's pread overlaps the current one's GPU work). **131-token prompt: 120.8 s → 75.4 s** (marginal 0.46 → 0.11 s/token; the gap widens with prompt length). Sequential replay retained behind `MFERENCE_INKLING_PREFILL=seq` as the correctness reference.

## Validation

- Four new Metal kernels parity-tested against CPU references; on the **real weights**: dense L0, MoE L2, global L5 attention, the head, and a 5-token deep-state composition — all cos ≥ 0.9999
- **Independent oracle**: layer-by-layer diff against upstream `inkling_mlx` on the same weights. This caught the final bug — dense layers' `mlp.global_scale` is BF16 while `gate.global_scale` is FP32; reading it as FP32 poisoned layers 0–1, invisible to the in-repo harness because engine and checker shared the accessor
- Fused QKV + constant-folded GEMV variants re-enabled after byte-identical A/B (both had been falsely suspected during a corrupt-shard incident)
- Conversion verified end to end; `--dry-run` install mode added
- **880 tests pass**

## Not in scope

- Vision/audio towers and MTP (text path only); context capped well below the advertised 1M (global-layer KV is 28.7 KB/token)

## Test plan

- [x] `swift test` — 880 tests, 143 suites green
- [x] Kernel CPU-parity suite (`InklingKernelTests`)
- [x] Real-model parity vs scalar reference (`InklingLayerParityTests`, env-gated)
- [x] Layer-by-layer + KLD + PPL vs upstream `inkling_mlx`
- [x] Greedy + sampled + chat-framed generation; chunked-vs-sequential prefill agreement

Note: All numbers are from MacBook M5 24GB

### CODE REVIEW

**Comment by NeelM0906** [comment]
Validation artifacts note: the source-safetensors oracle copy has been removed after the KLD/parity sweep completed (all shards were LFS-digest-verified before use; results in the PR body). The converted `.gturbo` is the runtime artifact. To re-run the reference oracle later, re-fetch the pinned revision `9d6e4720` and use `scratchpad/kld_ref.py`'s layer-freeing pattern from the session notes.

File: Sources/Mference/Runtime/Inference/RealForwardRunner.swift [hunk 1]
```diff
@@ -678,6 +790,12 @@ public final class RealForwardRunner: ChunkedPrefillRunner, ContextWindowReporti
 
     private func resetTransientState() {
         joinPendingSpeculation()
+        for state in inklingConvStates {
+            memset(state.k.contents(), 0, state.k.length)
+            memset(state.v.contents(), 0, state.v.length)
+            memset(state.attn.contents(), 0, state.attn.length)
+            memset(state.mlp.contents(), 0, state.mlp.length)
```
[the remark this answers is not in the document]
Comment (NeelM0906) [reply Sources/Mference/Runtime/Inference/RealForwardRunner.swift]: Confirmed — `prepareForContinuation` retains the KV cache but `resetTransientState` was wiping the conv history, so a resumed conversation's next token saw zero convolution state and diverged from a from-scratch prefill of the same tokens.

Fixed in e1db2e3: the zeroing moved to `reset()`, so conv history now lives and dies with the KV cache — the same lifecycle `gdnState` (Qwen's recurrent state) already follows. The conv buffers as left by the previous turn hold the last K-1 inputs of exactly the cached prefix, which is the correct continuation state; nothing needs reconstruction.

Regression test: `convStateSurvivesContinuationButNotReset` — sentinel-fills all four per-layer state buffers, asserts `reset()` clears them and `prepareForContinuation` preserves them.

File: Sources/MferenceRepack/Core/Verification/VerifiedInstallTool.swift [hunk 2]
```diff
@@ -156,6 +156,16 @@ public enum VerifiedInstallTool {
...
             }
+            // Leading dense-FFN layers carry no routed experts, so the writer
+            // emits an empty layout entry and no blob for them (Inkling's
+            // layers 0-1). There is nothing to size- or hash-check.
+            if layer.experts.isEmpty {
+                guard manifest.files["packed_experts/\(layer.file)"] == nil else {
+                    throw RepackError.configurationInvalid(
+                        detail: "packed_experts/\(layer.file) has no experts but is in the manifest")
+                }
+                continue
```
[the remark this answers is not in the document]
Comment (NeelM0906) [reply Sources/MferenceRepack/Core/Verification/VerifiedInstallTool.swift:181]: Confirmed — the empty-layout exception was unbounded, so an install missing a ROUTED layer's blob (with an empty `experts` array in layout.json) would pass `--verify-install` and earn a trusted receipt while `ManifestReader` rejects that same install at load.

Fixed in e1db2e3: the tool's manifest decode now includes `arch.numDenseLayers` and the exception is bounded to `layer.layer < numDenseLayers` (default 0, so pre-Inkling manifests get the old strict behavior). An empty routed layer now fails verification with an explicit message.

Regression test: `verifyRejectsEmptyLayoutOutsideDenseLayers` — crafts installs with the empty layout inside vs outside the dense prefix and asserts accept/reject respectively.
